### ISSUE TITLE
BUG: SciPy 1.7.x build failure on Cygwin

### ISSUE THREAD

**Issue by mkoeppe** [scipy/scipy#15296]
### Describe your issue.

SciPy 1.7.2 and 1.7.3 fail to build from source on Cygwin - https://trac.sagemath.org/ticket/33080

The failures come in through the vendored boost math library. as reported in https://github.com/boostorg/math/pull/727 and https://github.com/boostorg/math/pull/728

### Reproducing Code Example

```python
pip install scipy
```

### Error message

```shell
scipy/stats/_boost/src/beta_ufunc.cxx:3903:46:   required from here
  /opt/sage-4525b51c9c556e7d1bf7a4d90ee129c6b7f1d9a7/var/tmp/sage/build/scipy-1.7.3/src/scipy/_lib/boost/boost/math/tools/promotion.hpp:148:37: error: static assertion failed: Sorry, but this platform does not have sufficient long double support for the special functions to be reliably implemented.
    148 |          BOOST_STATIC_ASSERT_MSG((0 == ::boost::is_same<type, long double>::value), "Sorry, but this platform does not have sufficient long double support for the special functions to be reliably implemented.");
error: Command "g++ -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -ggdb -O2 -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong --param=ssp-buffer-size=4 -DOPENSSL_NO_SSL3=1 -fdebug-prefix-map=/pub/devel/python/python38/python38-3.8.12-1.x86_64/build=/usr/src/debug/python38-3.8.12-1 -fdebug-prefix-map=/pub/devel/python/python38/python38-3.8.12-1.x86_64/src/Python-3.8.12=/usr/src/debug/python38-3.8.12-1 -ggdb -O2 -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong --param=ssp-buffer-size=4 -DOPENSSL_NO_SSL3=1 -fdebug-prefix-map=/pub/devel/python/python38/python38-3.8.12-1.x86_64/build=/usr/src/debug/python38-3.8.12-1 -fdebug-prefix-map=/pub/devel/python/python38/python38-3.8.12-1.x86_64/src/Python-3.8.12=/usr/src/debug/python38-3.8.12-1 -D_GNU_SOURCE -DBOOST_MATH_DOMAIN_ERROR_POLICY=ignore_error -DBOOST_MATH_EVALUATION_ERROR_POLICY=user_error -DBOOST_MATH_PROMOTE_DOUBLE_POLICY=false -Iscipy/stats/_boost/include/ -Iscipy/stats/_boost/src/ -I/opt/sage-4525b51c9c556e7d1bf7a4d90ee129c6b7f1d9a7/lib/python3.8/site-packages/numpy/core/include -I/opt/sage-4525b51c9c556e7d1bf7a4d90ee129c6b7f1d9a7/var/tmp/sage/build/scipy-1.7.3/src/scipy/_lib/boost -I/opt/sage-4525b51c9c556e7d1bf7a4d90ee129c6b7f1d9a7/lib/python3.8/site-packages/numpy/core/include -Ibuild/src.cygwin-3.3.3-x86_64-3.8/numpy/distutils/include -I/opt/sage-4525b51c9c556e7d1bf7a4d90ee129c6b7f1d9a7/include -I/usr/include/python3.8 -c scipy/stats/_boost/src/beta_ufunc.cxx -o build/temp.cygwin-3.3.3-x86_64-3.8/scipy/stats/_boost/src/beta_ufunc.o -MMD -MF build/temp.cygwin-3.3.3-x86_64-3.8/scipy/stats/_boost/src/beta_ufunc.o.d -std=c++14 -msse -msse2 -msse3" failed with exit status 1
```

### SciPy/NumPy/Python version information

1.7.x

**Comment by mkoeppe** [comment]
https://github.com/boostorg/math/pull/729 (which consolidates https://github.com/boostorg/math/pull/727 and https://github.com/boostorg/math/pull/728) has just been merged into https://github.com/boostorg/math develop

**Comment by rgommers** [comment]
Upgrading https://github.com/scipy/boost-headers-only to the version with the fix is one option, just cherry-picking the fix into that repo and documenting that the next upgrade should be at least >=1.78 (check version) is another.

**Comment by mckib2** [comment]
> just cherry-picking the fix into that repo

This was going to my suggestion. @mkoeppe , would you be willing to make that PR into [scipy/boost-headers-only?](https://github.com/scipy/boost-headers-only )?

**Comment by mkoeppe** [comment]
Sure, that's now https://github.com/scipy/boost-headers-only/pull/5

**Comment by DWesl** [comment]
I still see this on 1.8.1 with the PyPI source package; is there some special way I'm supposed to get that source?

**Comment by rgommers** [comment]
Hmm indeed, the patch to `boost/math/tools/config.hpp` of https://github.com/scipy/boost-headers-only/pull/5 is not applied in the current version of the `boost` submodule in the main scipy repo. @mckib2 I don't know how this happened, but somehow that patch-application machinery did not work.

**Comment by mckib2** [comment]
> Hmm indeed, the patch to `boost/math/tools/config.hpp` of [scipy/boost-headers-only#5](https://github.com/scipy/boost-headers-only/pull/5) is not applied in the current version of the `boost` submodule in the main scipy repo. @mckib2 I don't know how this happened, but somehow that patch-application machinery did not work.

Hashes between latest boost-headers-only:main and scipy submodule match, but I suspect the patch file was added to the repo, but the headers never regenerated.  Regenerating now to test out that hypothesis

**Comment by mckib2** [comment]
https://github.com/scipy/boost-headers-only/pull/7 should at long last apply @mkoeppe's upstream patch.  We did take a circuitous route to get there, but I think maintaining the traceability of the generated headers is critical for long-term maintainability and well-worth the extra effort

**Comment by rgommers** [comment]
Thanks @mckib2, merged and opened gh-16370 to update this repo (needs backporting after merge so we include it in 1.9.0)

**Comment by rgommers** [comment]
Marked for 1.9.0, because the fix is marked for backport to `1.9.x`
